src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1140,6 +1230,123 @@ private void IssueBuildSubmissionToScheduler(BuildSubmission submission, bool al
             }
         }
 
+        private void ExecuteGraphBuildScheduler(BuildSubmission submission)
+        {
+            try
+            {
+                if (_shuttingDown)
+                {
+                    throw new BuildAbortedException();
+                }
+
+                var projectGraph = submission.BuildRequestData.ProjectGraph;
+                if (projectGraph == null)
+                {
+                    // TODO: Need to pass through other stuff? Look how RequestBuilder creates ProjectInstance
+                    projectGraph = new ProjectGraph(submission.BuildRequestData.ProjectGraphEntryPoints);
+                }
+
+                // TODO: Handle multiple entry points
+                var entryNode = projectGraph.EntryPointNodes.First();
+                var targetLists = projectGraph.GetTargetLists(submission.BuildRequestData.TargetNames);
+
+                var waitHandle = new AutoResetEvent(true);
+                var graphBuildStateLock = new object();
+
+                var blockedNodes = new HashSet<ProjectGraphNode>(projectGraph.ProjectNodes);
+                var finishedNodes = new HashSet<ProjectGraphNode>(projectGraph.ProjectNodes.Count);
+                var buildingNodes = new Dictionary<BuildSubmission, ProjectGraphNode>();
+                while (blockedNodes.Count > 0 || buildingNodes.Count > 0)
+                {
+                    waitHandle.WaitOne();
+
+                    lock (graphBuildStateLock)
+                    {
+                        var unblockedNodes = blockedNodes
+                            .Where(node => node.ProjectReferences.All(projectReference => finishedNodes.Contains(projectReference)))
+                            .ToList();
+                        foreach (var node in unblockedNodes)
+                        {
+                            var targetList = targetLists[node];
+                            if (targetList.Count == 0)
+                            {
+                                // An empty target list here means "no targets" instead of "default targets", so don't even build it.
+                                finishedNodes.Add(node);
+                                blockedNodes.Remove(node);
+
+                                waitHandle.Set();
+
+                                continue;
+                            }
+
+                            var request = new BuildRequestData(node.ProjectInstance, targetList.ToArray());
+
+                            // TODO Tack onto the existing submission instead of pending a whole new submission for every node
+                            // Among other things, this makes BuildParameters.DetailedSummary produce a summary for each node, which is not desirable.
```

Comment:
Do you have a clear picture of how to do this? Detailed summary is pretty important, I think.

Reply:
No clear picture unfortunately. I can elaborate an idea of what I think should happen though.